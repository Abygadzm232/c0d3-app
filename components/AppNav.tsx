import React, { useEffect, useState } from 'react'
import NavLink, { NavLinkProps } from './NavLink'
import { useRouter } from 'next/router'
import { useGetAppQuery, GetAppQuery } from '../graphql'
import Navbar from 'react-bootstrap/Navbar'
import { Container, Nav } from 'react-bootstrap'
import ProfileDropdownMenu from './ProfileDropdownMenu'
import styles from '../scss/appNav.module.scss'
import _ from 'lodash'

type AuthLinkProps = {
  session: any
}

type NavItem = NavLinkProps & { name: string }

const navItems: NavItem[] = [
  { path: '/curriculum', name: 'Curriculum' },
  {
    path: 'https://github.com/garageScript/c0d3-app',
    name: 'Repo',
    external: true
  },
  {
    path: '/docs/learn',
    name: 'Learn'
  },
  {
    path: 'https://discord.gg/c0d3',
    name: 'Community',
    external: true
  }
]

const NavBar: React.FC<AuthLinkProps> = ({}) => {
  const router = useRouter()
  const location = router.asPath

  return (
    <>
      {navItems.map(button => (
        <NavLink
          {...button}
          className={`${styles['nav-item']} ${styles['nav-link']} px-3`}
          key={button.name}
          activePath={location.includes(button.path)}
        >
          {button.name}
        </NavLink>
      ))}
    </>
  )
}

const NotLoggedInAuthNav = () => (
  <div className={`${styles['nav-buttons']}`}>
    <NavLink path="/login" className={`btn m-2 me-lg-3 ${styles['login-btn']}`}>
      Login
    </NavLink>
    <NavLink
      path="/signup"
      className={`btn m-2 me-lg-3 ${styles['signup-btn']}`}
    >
      Signup
    </NavLink>
  </div>
)

const AppNav: React.FC<{}> = () => {
  const [session, setSession] = useState<GetAppQuery['session']>()
  const isAdmin = _.get(session, 'user.isAdmin', false)

  const { data, loading } = useGetAppQuery()

  useEffect(() => {
    if (data && data.session) {
      setSession(data.session)
    }
  }, [data])

  const renderButtons = () => {
    if (loading) return <></>

    if (!session || _.get(session, 'user.username', null) === null) {
      return <NotLoggedInAuthNav />
    }

    // TODO: replace with typing
    const username = _.get(session, 'user.username', '')

    return <ProfileDropdownMenu username={username} isAdmin={isAdmin} />
  }

  return (
    <Navbar
      id="appNav"
      className={`${styles['navbar']}`}
      expand="lg"
      bg="white"
    >
      <Container>
        <Navbar.Brand href="/">
          <div className={`${styles['navbar-brand']} text-primary fw-bold`}>
            C0D3
            {isAdmin ? (
              <span className={`${styles['isAdminBadge']} badge`}> ADMIN </span>
            ) : (
              ''
            )}
          </div>
        </Navbar.Brand>
        <Navbar.Toggle aria-controls="basic-navbar-nav" />
        <Navbar.Collapse id="basic-navbar-nav" className="text-center">
          <Nav className="m-auto">
            <NavBar session={session} />
          </Nav>
          {renderButtons()}
        </Navbar.Collapse>
      </Container>
    </Navbar>
  )
}

export default AppNav
