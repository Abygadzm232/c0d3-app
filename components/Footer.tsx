import React from 'react'

type Props = {
  footerType?: string
}

const Footer: React.FC<Props> = ({ footerType }) => {
  const footerClass = footerType
    ? `font-weight-light text-center ${footerType}`
    : `mt-5 font-weight-light text-center pb-5 text-secondary`
  return (
    <footer className={footerClass}>
      &copy; Copyright 2020 GarageScript
      <br />
      All Rights Reserved
    </footer>
  )
}

export default Footer
