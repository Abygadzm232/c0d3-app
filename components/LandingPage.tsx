import * as React from 'react'

const LandingPage: React.FC = () => {
  return (
    <React.Fragment>
      <div className="container text-center pb-5">
        <div className="row justify-center mt-5">
          <div className="col-md-6 offset-md-3">
            <img src="header-01.svg" alt="Hero Image"></img>
          </div>
        </div>
        <div className="row justify-center mt-5">
          <div className="col-md-8 offset-md-2">
            <h1 style={{ fontWeight: 900, fontSize: '3.8rem' }}>
              Become a Software Engineer the Hard Way
            </h1>
          </div>
        </div>
        <div className="row justify-center mt-4">
          <div className="col-md-10 offset-md-1">
            <h4 style={{ lineHeight: 1.5 }} className="font-weight-light">
              Learn the right foundations you need to become a full stack
              software engineer. Our curriculum takes no shortcuts and
              effectively trains students with no technical background to become
              effective software engineers.
            </h4>
          </div>
        </div>
        <div className="row mt-5">
          <div className="col-md-6 offset-md-3 mb-5">
            <a
              href="https://c0d3.com/book"
              className="btn btn-primary py-3 px-5"
            >
              Get Started
            </a>
          </div>
        </div>
        <div className="row mt-5">
          <div className="col-md-6 offset-md-3">
            <h1 className="display-4 font-weight-bold" id="learning">
              Learning Process
            </h1>
          </div>
        </div>
        <div className="row mt-5">
          <div className="col-md-10 offset-md-1">
            <h4 className="font-weight-light" style={{ lineHeight: 1.5 }}>
              Our learning process is interactive and follows the same practice
              as a well functioning engineering team. From the first line of
              code you write, you will be code reviewed by engineers to ensure
              you are always following industry best practices.
            </h4>
          </div>
        </div>
        <div className="row mt-5 align-items-start mb-5">
          <div className="col-md-4">
            <img
              src="choose-01.svg"
              alt="Curriculum"
              style={{ height: '170px' }}
            ></img>
            <h4 className="mt-3 font-weight-bold">Modern Curriculum</h4>
            <p className="mt-4 font-weight-light">
              First the curriculum will help you achieve mastery in JavaScript.
              Using JavaScript, you will learn about different branches of
              software engineering: websites, mobile applications, servers, and
              machine learning.
            </p>
          </div>
          <div className="col-md-4">
            <img
              src="choose-02.svg"
              alt="Support"
              style={{ height: '170px' }}
            ></img>
            <h4 className="mt-3 font-weight-bold">Support</h4>
            <p className="mt-4 font-weight-light">
              One of the most important skills in a challenging workplace is
              knowing how to ask for help. Our curriculum is intentionally
              difficult and we have dedicated mentors to answer your questions
              and guide you how to communicate effectively.
            </p>
          </div>
          <div className="col-md-4">
            <img
              src="choose-03.svg"
              alt="Curriculum Image"
              style={{ height: '170px' }}
            />
            <h4 className="mt-3 font-weight-bold">Work Experience</h4>
            <p className="mt-4 font-weight-light">
              Once you have completed our curriculum and learned to communicate
              effectively, you will be invited to contribute to this open-source
              project to improve the learning experience for future students.
            </p>
          </div>
        </div>
      </div>
      <div
        className="container-fluid text-center py-5"
        style={{ background: 'rgb(84,64,216)', color: 'white' }}
      >
        <div className="container mt-5">
          <h1 id="journey" className="mt-5 display-4 font-weight-bold">
            Your Learning Journey
          </h1>
          <h4
            className="mt-5 font-weight-light col-md-10 offset-md-1 pb-5"
            style={{ lineHeight: 1.5 }}
          >
            Our learning process is interactive and follows the same practice as
            a well functioning engineering team. From the first line of code you
            write, you will be code reviewed by engineers to ensure you are
            always following industry best practice
          </h4>
          <div className="row mt-5 pb-5">
            <div className="col-md-6 order-md-2">
              <img src="journey-01.svg" className="mb-3"></img>
            </div>
            <div className="col-md-6 order-md-1">
              <h2 className="font-weight-bold">Creating an account</h2>
              <p className="mt-3 font-weight-light">
                After creating an account, you will gain access to our code
                reviewed curriculum and our community chat. You will be able to
                setup c0d3.com CLI which will allow you to submit code for
                reivew. The CLI allows you to code in Javascript from any code
                editor of your choice.
              </p>
            </div>
          </div>
          <div className="row mt-5 pb-5">
            <div className="col-md-6">
              <img src="journey-02.svg" className="mb-3" />
            </div>
            <div className="col-md-6">
              <h2 className="font-weight-bold">Curriculum</h2>
              <p className="mt-3 font-weight-light">
                Each lesson in our curriculum is designed to be hands on and you
                will learn by working through plenty of exercises. There will be
                video lectures to guide you along as well as lecture notes. In
                the interactive lessons where you write server code and build
                web applications, you will be able to create your own subdomains
                and your code is immediately available on c0d3.com around the
                world. If you get stuck, you can ask for help in our chat
                channel, where students who passed the lessons would be helping
                you along.
              </p>
            </div>
          </div>
          <div className="row mt-5 pb-5">
            <div className="col-md-6 order-md-2">
              <img src="journey-03.svg" className="mb-3" />
            </div>
            <div className="col-md-6 order-md-1">
              <h2 className="font-weight-bold">Conducting Code Reviews</h2>
              <p className="mt-3 font-weight-light">
                When you finish your lesson, you will have the opportunity to
                help new students who are starting the lesson. By code reviewing
                students to the same standard that you were code reviewed, you
                help ensure that new students are writing good code and becoming
                better engineers. At the same time, conducting regular code
                reviews helps you improve your ability to read/debug code as
                well as understand the lesson concepts at a deeper level.
              </p>
            </div>
          </div>
          <div className="row mt-5 pb-5">
            <div className="col-md-6">
              <img src="journey-04.svg" className="mb-3" />
            </div>
            <div className="col-md-6">
              <h2 className="font-weight-bold">Work Experience</h2>
              <p className="mt-3 font-weight-light">
                After completing our curriculum, you will gain access to our
                internal code repository that powers c0d3.com and have the
                opportunity to work with our engineers and experience the entire
                workflow of a full stack software engineer. There will be bugs
                and feature list on our issue board for you to contribute to and
                you will be held to the same expectations as any respected
                engineer: high quality code and proper testing. Every feature
                contribution will be documented and recogized on our
                contribution list and you would be able to list your
                contributions as work experience on your resume.
              </p>
            </div>
          </div>
        </div>
      </div>
    </React.Fragment>
  )
}

export default LandingPage
