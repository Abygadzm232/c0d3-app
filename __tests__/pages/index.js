jest.mock('../../pages/curriculum')
jest.mock('../../components/AppNav')
jest.mock('../../components/LandingPage')
jest.mock('../../components/Footer')
import React from 'react'
import { render } from '@testing-library/react'
import Curriculum from '../../pages/curriculum'
import IndexPage from '../../pages/index'
import LandingPage from '../../components/LandingPage'
import Footer from '../../components/Footer'
import AppNav from '../../components/AppNav'
import SessionContext from '../../helpers/contexts/session'

describe('Index Page', () => {
  Footer.mockReturnValue(<h1>Footer</h1>)
  AppNav.mockReturnValue(<h1>AppNav</h1>)
  LandingPage.mockReturnValue(<h1>LandingPage</h1>)
  Curriculum.mockReturnValue(<h1>Hello Curriculum</h1>)

  test('Should render curriculum if session is identified', async () => {
    const session = {
      session: {
        user: {
          username: 'testing'
        }
      }
    }
    const tree = (
      <SessionContext.Provider value={session}>
        <IndexPage />
      </SessionContext.Provider>
    )

    const { getByText } = render(tree)
    getByText('Hello Curriculum')
  })

  test('Should render landing page if user is not identified', async () => {
    const session = { session: null }
    const tree = (
      <SessionContext.Provider value={session}>
        <IndexPage />
      </SessionContext.Provider>
    )

    const { getByText } = render(tree)
    getByText('LandingPage')
    getByText('Footer')
  })
  test('Should not render while page is loading', async () => {
    // SessionContext default value is { data: { success: false } }
    const { container } = render(<IndexPage />)
    expect(container).toMatchSnapshot()
  })
})
