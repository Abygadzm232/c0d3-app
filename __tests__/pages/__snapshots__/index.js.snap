// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index Page Should not render while page is loading 1`] = `
<div>
  <h1>
    AppNav
  </h1>
  <h1>
    LandingPage
  </h1>
  <h1>
    Footer
  </h1>
</div>
`;
